contracts/YearnOTC/YearnOTCPoolTradeable.sol

```diff
@@ -0,0 +1,103 @@
+
+  function setSwapperRegistry(address _swapperRegistry) external;
+
+  function claim(address _token, uint256 _amount) external;
+
+  function takeOffer(
+    address _offeredTokenToPool,
+    address _wantedTokenFromPool,
+    uint256 _maxOfferedAmount
+  ) external returns (uint256 _tookFromPool, uint256 _tookFromSwapper);
+}
+
+abstract contract YearnOTCPoolTradeable is IYearnOTCPoolTradeable, YearnOTCPoolDesk {
+  using SafeERC20 for IERC20;
+
+  address public override swapperRegistry;
+  mapping(address => uint256) public override swappedAvailable;
+
+  constructor(address _swapperRegistry) {
+    swapperRegistry = _swapperRegistry;
+  }
+
+  function _setSwapperRegistry(address _swapperRegistry) internal {
+    require(_swapperRegistry != address(0), 'YearnOTCPool: zero address');
+    swapperRegistry = _swapperRegistry;
+    emit SwapperRegistrySet(_swapperRegistry);
+  }
+
+  modifier onlyRegisteredSwapper {
+    require(SwapperRegistry(swapperRegistry).isSwapper(msg.sender), 'YearnOTCPool: not a registered swapper');
+    _;
+  }
+
+  function _claim(
+    address _receiver,
+    address _token,
+    uint256 _amountToClaim
+  ) internal {
+    require(_receiver != address(0), 'YearnOTCPool: receiver should not be zero');
+    require(_token != address(0), 'YearnOTCPool: token should not be zero');
+    require(_amountToClaim > 0, 'YearnOTCPool: should provide more than zero');
+    _reduceSwappedAvailable(_token, _amountToClaim);
+    IERC20(_token).safeTransfer(_receiver, _amountToClaim);
+    emit Claimed(_receiver, _token, _amountToClaim);
+  }
+
+  function _reduceSwappedAvailable(address _token, uint256 _amountToClaim) internal {
+    require(_amountToClaim >= swappedAvailable[_token], 'YearnOTCPool: swapped not available');
+    swappedAvailable[_token] -= _amountToClaim;
+  }
+
+  // TODO: Discuss if this logic should be here, or should we make this calcs inside the OTC swapper
+  // and trust that contract (philosophical discussion)
+  function _performTradeOnSwapper(
+    address _swapper,
+    address _offeredTokenToPool,
+    address _wantedTokenFromPool,
+    uint256 _maxOfferedAmount
+  ) internal returns (uint256 _tookFromPool, uint256 _tookFromSwapper) {
+    if (availableFor[_wantedTokenFromPool][_offeredTokenToPool] == 0) return (0, 0);
+    uint256 _maxWantedFromOffered = IYearnOTCSwapper(_swapper).getTotalAmountOut(_offeredTokenToPool, _wantedTokenFromPool, _maxOfferedAmount);
+    _tookFromPool = Math.min(availableFor[_wantedTokenFromPool][_offeredTokenToPool], _maxWantedFromOffered);
+    _tookFromSwapper = IYearnOTCSwapper(_swapper).getTotalAmountOut(_wantedTokenFromPool, _offeredTokenToPool, _tookFromPool);
+    IERC20(_offeredTokenToPool).safeTransferFrom(_swapper, address(this), _tookFromSwapper);
+    _performTrade(_offeredTokenToPool, _wantedTokenFromPool, _tookFromPool, _tookFromSwapper);
+    emit TradePerformed(_swapper, _offeredTokenToPool, _wantedTokenFromPool, _tookFromPool, _tookFromSwapper);
+  }
```

Comment:
@lbertenasco all the `getTotalAmount` stuff could be easily calculated FROM the swapper itself and sent to this function and it would save a shit ton of gas probably. But this would mean **trusting** that the swapper is a good guy / good swapper / not malicious.

At the same time, we are already trusting that ... Since if `getTotalAmount` is tinkered in a swapper that would mean rektening. 

So ... either way we have to trust, the question is how much. It would be much appreciated some input here.